packages/nocodb/src/lib/meta/api/dataApis/bulkDataAliasApis.ts

```diff
@@ -1,75 +1,77 @@
 import { Request, Response, Router } from 'express';
+import { BaseModelSqlv2 } from '../../../db/sql-data-mapper/lib/sql/BaseModelSqlv2';
 import Model from '../../../models/Model';
 import Base from '../../../models/Base';
 import NcConnectionMgrv2 from '../../../utils/common/NcConnectionMgrv2';
 import ncMetaAclMw from '../../helpers/ncMetaAclMw';
 import { getViewAndModelFromRequestByAliasOrId } from './helpers';
 import apiMetrics from '../../helpers/apiMetrics';
 
-async function bulkDataInsert(req: Request, res: Response) {
+async function getModelAndBase(req: Request) {
   const { model, view } = await getViewAndModelFromRequestByAliasOrId(req);
 
   const base = await Base.get(model.base_id);
+  return { model, view, base };
+}
 
+async function executeBulkOperation(
+  req: Request,
+  res: Response,
+  operation:
+    | 'bulkInsert'
+    | 'bulkUpdate'
+    | 'bulkUpdateAll'
+    | 'bulkDelete'
+    | 'bulkDeleteAll',
+  options:
+    | Parameters<typeof BaseModelSqlv2.prototype.bulkInsert>
+    | Parameters<typeof BaseModelSqlv2.prototype.bulkUpdate>
+    | Parameters<typeof BaseModelSqlv2.prototype.bulkDelete>
+    | Parameters<typeof BaseModelSqlv2.prototype.bulkUpdateAll>
+    | Parameters<typeof BaseModelSqlv2.prototype.bulkDeleteAll>
+) {
```

Comment:
since you are revising options types as well, I think it could be further simplified as 

```ts
async function executeBulkOperation<
  operation extends
    | 'bulkInsert'
    | 'bulkUpdate'
    | 'bulkUpdateAll'
    | 'bulkDelete'
    | 'bulkDeleteAll'
>(
  req: Request,
  res: Response,
  operation,
  options: Parameters<typeof BaseModelSqlv2.prototype[operation]>
) {
```

Reply:
Yes, I had already thought of that and tested it that way. Somehow I decided against it. Probably because I thought you would have to pass the template variable (redundantly) when calling the function. But it seems to be working (as it currently is). Good that you exchange ideas in the community :)